# Kapitel 3 - Regression

## Faktafrågor

<font color='orange'>1. Vad kännetecknar regressionsproblem? Ge några exempel på tillämpningsområden. </font>

Svar: Regressionsproblem handlar då om att prediktera en kontinuerlig och beroende variabel med hjälp av en oberoende variabel. Exempel: Bilvärdering, försäkringar och Lönenivå

<font color='orange'>2. Förklara utvärderingsmåtten *RMSE*, *MSE* och *MAE*.</font>

Svar:
- RMSE: Root mean squared error, används mest i samband med att regressionsmodeller utvärderas. Hårdast mod extrema värden då man drar roten ur på talen.
- MSE: Mean squared error, används om vi bara är ute efter rangordningen och inte talet. detta då det är mer snabbare pga att kvadratroten inte används
- MAE: Mean absolute error, tar fram ett absolutbelopp. Detta är då ett tal som räknar totala missen. exempel: |2| = 2 och |-2| = 2.

<font color='orange'>3. Om vi ska rangordna olika modeller, spelar det någon roll om *RMSE* eller *MSE* används? Varför? </font>

Svar: MSE används om vi bara är ute efter rangordningen och inte talet. detta då det är mer snabbare pga att kvadratroten inte används. Annars används RMSE då den skapar bättre jämförelsetal

<font color='orange'>4. Förklara mycket översiktligt vad *gradient descent* är. </font>

Svar: Gradient descent är en optimeringsalgoritm som används för att hitta en funktions minimum eller maximum. 

<font color='orange'>5. Vad är *the bias variance trade-off*? Varför är mer komplexa modeller inte alltid bättre? </font>

Svar: den säger att felet en modell gör när den ska prediktera en ny. den osedda datan kan då delas pp i 3 delar. Bias, variance, irreducerbart fel.

Det är för att den kan då öka dens varians men minska bias. Vilket kan ge negativa följder.

<font color='orange'>6. Några vanligt förekommande modeller för regressionsproblem är enligt nedan. Förklara översiktligt hur respektive modell fungerar. Läs även igenom respektive modells dokumentation, notera att du inte behöver förstå alla detaljer från dokumentationen men det är bra att ha läst igenom den. 

a) Linjär regression - [https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html)

b) *Ridge* regression - [https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ridge_regression.html](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ridge_regression.html)

c) *Lasso* regression - [https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Lasso.html](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Lasso.html)

d) *Elastic net* - [https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ElasticNet.html](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ElasticNet.html)

e) *Support vector machines* - [https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVR.html](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVR.html)

f) Beslutsträd - [https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeRegressor.html](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeRegressor.html)

g) *Ensemble learning* - [https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.VotingRegressor.html](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.VotingRegressor.html) och [https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingRegressor.html](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingRegressor.html)

h) *Random forest* - [https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html)</font>

Svar:

a) Linjär regression anpassar en linjär modell med koefficienter för att minska kvadratsumman av reidualerna mellan de observerade målvärdenai datamängden och de målvärden som förutsägs av den linjära approximationen.

b) Ridge regression är en regulariserad varit av linjära regressionsmodeller.

c) Lasso regression: Använder L1-regularisering (absolutbeloppet av vikterna) för att sätta oviktiga variabler till exakt noll (automatisk variabelselektion).

d) Elastic Net: Kombinerar både L1- (Lasso) och L2-regularisering (Ridge) i samma modell.

e) Support Vector Machines (SVR): Skapar en felemotståndig marginal kring prediktionslinjen och använder kernel-tricket för icke-linjära samband.

f) Beslutsträd: Delar upp datan sekventiellt i villkorsbaserade grenar för att minimera variansen i lövnoderna.

g) Ensemble Learning: Slår samman förutsägelser från flera modeller (t.ex. Voting eller Bagging) för att minska felet.

h) Random Forest: Kombinerar många beslutsträd där varje noduppdelning endast får använda en slumpmässig delmängd av variablerna.

<font color='orange'>7. Vad menas med *white box* modeller och *black box* modeller? </font>

svar:

White box-modeller: Transparenta och lätta att tolka. Det går enkelt att förstå och förklara exakt hur besluten fattas 

Black box-modeller: Komplexa och svårgenomskådliga. De har ofta hög träffsäkerhet men det är svårt att förklara *varför* en viss prediktion görs 

<font color='orange'>8. Vad är skillnaden mellan *bagging* och *pasting*?</font>

Svar:
Skillnaden är att bagging gör urval med återläggning (bootstrap=True), vilket gör att samma datapunkt kan väljas flera gånger i ett deldataset, medan pasting gör urval utan återläggning (bootstrap=False) så att varje datapunkt endast kan väljas en gång.

## Resonemangfrågor

<font color='orange'>9. Förklara hur man kan tolka figur 3.1 på sidan 113. </font>

svar:
figuren ger en schematisk bild över den enkla linjära regressionsmodellen och tolkas utifrån följande fem komponenter.
-Svarta punkterna
-Blå linjen
-Interceptet
-Lutningen
-Röda vertikala linjerna

<font color='orange'>10. Förklara hur man kan tolka figur 3.13 på sidan 140. Hur hänger den ihop med figur 3.14 på sidan 141? </font>

svar:
Noder: Visar villkoret för uppdelning, antal observationer (samples) och felet (MSE).

Förgreningar: Styr observationer vidare beroende på om villkoret är sant eller falskt.

Lövnoder: Innehåller modellens slutgiltiga prediktion, vilket är medelvärdet av alla träningsdatapunkter i den noden.

Figur 3.13 visar beslutsträdets logiska flöde av villkor, vilka i Figur 3.14 utgör de vinkelräta delningslinjer som delar upp datarummet i motsvarande avgränsade regioner

<font color='orange'>11. Vi har lärt oss utvärderingsmåtten *RMSE*, *MSE* och *MAE*. Ett annat utvärderingsmått är det som benämns för determinationskoefficienten eller $R^2$. Förklara vad det är för mått. Du kan läsa dokumentationen om det på nedanstående två länkar. 
* [https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)

* [https://scikit-learn.org/stable/modules/model_evaluation.html#r2-score](https://scikit-learn.org/stable/modules/model_evaluation.html#r2-score)</font>

svar:

Determinationskoefficienten mäter hur stor andel av variansen i målvariabeln som förklaras av modellens oberoende variabler. Det bästa möjliga värdet är 1.0 för en perfekt modell, medan 0.0 innebär att modellen inte presterar bättre än att gissa på medelvärdet (och negativa värden indikerar ännu sämre prediktioner) Till skillnad från RMSE och MAE ger R^2 ett skalfritt mått på modellens anpassningsförmåga (goodness of fit) som gör det enkelt att jämföra prestanda mellan olika dataset.

## Koduppgifter

<font color='orange'>12. Gå igenom samtliga kodexempel i kapitlet och skriv gärna av koden manuellt. Det är även bra att experimentera genom att ändra vissa delar av koden och läsa dokumentationen. </font>

<font color='orange'>13. Datasetet "hr_employee_data.xlsx" innehåller data kopplat till ett företags anställda. Gör en *EDA* och tillhörande analys, föreställ dig att ledningsgruppen på ett företag ska ta del av analysen. I denna uppgift ska ingen ML-modellering göras (det kommer göras i nästa kapitel). Syftet är dels att träna på att genomföra analyser dels att demonstrera hur mycket man kan göra även utan ML-modellering.  </font>

Svar:
EDA:n görs i flera steg: datakvalitet, övergripande personalomsättning (attrition), och vad som skiljer anställda som slutat (left=1) från de som stannat. Uppgiften avslutas med en sammanfattning riktad till ledningsgruppen. Ingen ML-modellering görs här, det kommer i nästa kapitel.

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)

df = pd.read_excel('dataset/hr_employee_data.xlsx')
print(df.shape)
df.head()

(14999, 11)


,Emp_Id,satisfaction_level,last_evaluation,number_project,average_montly_hours,time_spend_company,Work_accident,left,promotion_last_5years,Department,salary
0,IND02438,0.38,0.53,2,157,3,0,1,0,sales,low
1,IND28133,0.80,0.86,5,262,6,0,1,0,sales,medium
2,IND07164,0.11,0.88,7,272,4,0,1,0,sales,medium
3,IND30478,0.72,0.87,5,223,5,0,1,0,sales,low
4,IND24003,0.37,0.52,2,159,3,0,1,0,sales,low


Datakvalitet:
Datatyperna ser rimliga ut och det saknas inga värden. Emp_Id är unikt för varje rad, men om Emp_Id exkluderas finns det ca 3000 rader som är identiska på övriga kolumner. Det kan bero på att flera anställda råkar ha exakt samma värden (rimligt när flera kolumner är avrundade/diskreta), men det är även en känd svaghet i det här datasetet och något man bör flagga för ledningsgruppen som en osäkerhetsfaktor i analysen.

In [2]:
print(df.dtypes)
print()
print('Saknade värden per kolumn:')
print(df.isna().sum())
print()
print('Antal dubblettrader (exkl. Emp_Id):', df.drop(columns='Emp_Id').duplicated().sum())

Emp_Id                       str
satisfaction_level       float64
last_evaluation          float64
number_project             int64
average_montly_hours       int64
time_spend_company         int64
Work_accident              int64
left                       int64
promotion_last_5years      int64
Department                   str
salary                       str
dtype: object

Saknade värden per kolumn:
Emp_Id                   0
satisfaction_level       0
last_evaluation          0
number_project           0
average_montly_hours     0
time_spend_company       0
Work_accident            0
left                     0
promotion_last_5years    0
Department               0
salary                   0
dtype: int64

Antal dubblettrader (exkl. Emp_Id): 3008


Övergripande personalomsättning:
Knappt en fjärdedel av de anställda i datasetet har slutat, vilket är en betydande andel och en rimlig utgångspunkt för resten av analysen.

In [3]:
churn_rate = df['left'].mean()
print(f'Andel anställda som slutat: {churn_rate:.1%}')
df['left'].value_counts()

Andel anställda som slutat: 23.8%


left
0    11428
1     3571
Name: count, dtype: int64

Vad skiljer de som slutat från de som stannat?
De som slutat har i genomsnitt klart lägre satisfaction_level, något fler projekt, fler arbetstimmar per månad och längre anställningstid. last_evaluation skiljer sig knappt alls mellan grupperna, vilket redan här antyder att det inte enbart är lågpresterande medarbetare som slutar.

In [4]:
df.groupby('left')[
    ['satisfaction_level', 'last_evaluation', 'number_project', 'average_montly_hours', 'time_spend_company']
].mean().T

left,0,1
satisfaction_level,0.666810,0.440098
last_evaluation,0.715473,0.718113
number_project,3.786664,3.855503
average_montly_hours,199.060203,207.419210
time_spend_company,3.380032,3.876505


Korrelation mot left visar att satisfaction_level är den enskilt starkast kopplade variabeln (negativ korrelation), följt av tidsspenderad i företaget (positiv) och arbetsplatsolycka (negativ, se nedan).

In [5]:
df.select_dtypes('number').corr()['left'].sort_values(ascending=False)

left                     1.000000
time_spend_company       0.144822
average_montly_hours     0.071287
number_project           0.023787
last_evaluation          0.006567
promotion_last_5years   -0.061788
Work_accident           -0.154622
satisfaction_level      -0.388375
Name: left, dtype: float64

Ett medelvärde döljer dock ett viktigt mönster. Om vi delar upp enbart de som slutat i nöjdhetsgrupper syns tydligt att de flesta antingen är väldigt missnöjda eller relativt nöjda, medan mellangruppen (0.5–0.7) är nästan tom. Det tyder på minst två olika typer av medarbetare som slutar: dels missnöjda medarbetare som troligen är underpresterande eller feltillsatta, dels nöjda/högpresterande medarbetare som ändå väljer att sluta, sannolikt på grund av överarbete eller att de blir headhuntade. Den senare gruppen fångas inte upp av en enkel nöjdhetsundersökning.

In [6]:
left_df = df[df['left'] == 1]
bins = pd.cut(left_df['satisfaction_level'], bins=[0, 0.2, 0.5, 0.7, 1.0])
bins.value_counts().sort_index()

satisfaction_level
(0.0, 0.2]     924
(0.2, 0.5]    1626
(0.5, 0.7]      88
(0.7, 1.0]     933
Name: count, dtype: int64

Lönenivå har ett starkt samband med personalomsättning: andelen som slutar är nästan fem gånger så hög bland lågavlönade som bland högavlönade. Skillnaderna mellan avdelningar är mindre men HR och accounting sticker ut med högre andel avhopp, medan management och R&D har lägst.

In [7]:
print('Andel som slutat per lönenivå:')
print(df.groupby('salary')['left'].mean().sort_values(ascending=False))
print()
print('Andel som slutat per avdelning:')
print(df.groupby('Department')['left'].mean().sort_values(ascending=False))

Andel som slutat per lönenivå:
salary
low       0.296884
medium    0.204313
high      0.066289
Name: left, dtype: float64

Andel som slutat per avdelning:
Department
hr             0.290934
accounting     0.265971
technical      0.256250
support        0.248991
sales          0.244928
marketing      0.236597
IT             0.222494
product_mng    0.219512
RandD          0.153748
management     0.144444
Name: left, dtype: float64


Anställda som blivit befordrade de senaste 5 åren slutar betydligt mer sällan, vilket tyder på att karriärmöjligheter är en viktig retentionsfaktor - dock är det bara ca 2 % av alla anställda som blivit befordrade. Att ha haft en arbetsplatsolycka hänger också ihop med lägre andel avhopp, men det bör tolkas försiktigt - det är rimligare att det återspeglar mer erfarna/etablerade medarbetare än att olyckan i sig skulle minska benägenheten att sluta.

In [8]:
print('Andel som slutat, uppdelat på om man haft en arbetsplatsolycka:')
print(df.groupby('Work_accident')['left'].mean())
print()
print('Andel som slutat, uppdelat på om man blivit befordrad senaste 5 åren:')
print(df.groupby('promotion_last_5years')['left'].mean())

Andel som slutat, uppdelat på om man haft en arbetsplatsolycka:
Work_accident
0    0.265160
1    0.077916
Name: left, dtype: float64

Andel som slutat, uppdelat på om man blivit befordrad senaste 5 åren:
promotion_last_5years
0    0.241962
1    0.059561
Name: left, dtype: float64


Antal projekt har ett tydligt U-format samband med att sluta: medarbetare med bara 2 projekt slutar i mycket hög utsträckning (troligen understimulerade/lågt engagerade), medan de med 6–7 projekt slutar ännu oftare (troligen överarbetade). Andelen som slutar är som lägst för de med 3–4 projekt, vilket verkar vara en "sweet spot" för arbetsbelastning.

In [9]:
df.groupby('number_project')['left'].agg(antal='size', andel_slutat='mean')

,antal,andel_slutat
number_project,,
2,2388,0.656198
3,4055,0.017756
4,4365,0.093700
5,2761,0.221659
6,1174,0.557922
7,256,1.000000


Sammanfattning för ledningsgruppen:
- Personalomsättningen är hög: knappt 24 % av de anställda i datasetet har slutat.
- Det finns inte bara en typ av avhopp. En stor grupp som slutar är tydligt missnöjd (satisfaction_level under 0.2), men en nästan lika stor grupp är faktiskt relativt nöjd (över 0.7) och slutar ändå - troligen på grund av överarbete eller att de blir headhuntade. Den gruppen fångas inte upp av vanliga nöjdhetsmätningar.
- Arbetsbelastning hänger ihop med avhopp i ett U-format mönster: både för få (2) och för många (6-7) projekt är kopplat till hög andel avhopp, medan 3-4 projekt är som lägst.
- Lön spelar stor roll: lågavlönade slutar nästan fem gånger så ofta som högavlönade (ca 30 % mot ca 7 %).
- Befordran är en stark retentionsfaktor (andelen som slutar sjunker från ca 24 % till ca 6 %), men mycket få anställda (ca 2 %) blir befordrade under en femårsperiod.
- Rekommendationer: se över arbetsbelastningen så färre medarbetare hamnar i ytterligheterna, bredda karriär- och befordringsmöjligheterna, granska lönenivåerna i botten av skalan, och håll extra koll på nöjda högpresterande medarbetare eftersom de inte flaggas av traditionella missnöjdhetssignaler.
- Notera att ca 3000 rader är identiska förutom Emp_Id, vilket är en osäkerhetsfaktor i underlaget och bör undersökas närmare innan analysen används för beslut.
- Ingen ML-modell har tränats i denna uppgift, men mönstren ovan ger redan ett konkret underlag att agera på inför nästa kapitels modellering.

<font color='orange'>14. Nedan ser vi en början till modellering av ett dataset kopplat till diabetes. Läs på vad datasetet innebär och genomför ett komplett ML-flöde. Börja med att göra ett grundläggande flöde som fungerar, därefter kan du om du önskar försöka förbättra resultatet. 
```{python}
#| eval: false

from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split

# This code is merely executed to see the description of the data in a smooth way
data = load_diabetes()

print(data.DESCR)

# Storing/loading the data the way it will be used
X, y = load_diabetes(return_X_y=True, as_frame=True)
print(X.info())
print(y.info())

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
```
</font> 

<font color='orange'>15. I denna uppgift ska vi arbeta med datasetet "car_price_dataset.csv". 
a) Gör ett komplett ML flöde där den beroende variabeln, $y$, är `price`. Vi vill alltså skapa en modell som kan prediktera en bils pris. Ett tips är att först göra ett enkelt flöde som fungerar, har du därefter tid kan du försöka förbättra resultaten. 

b) Gör en applikation med hjälp av *Streamlit* där användaren kan mata in olika uppgifter om en bil och därefter få ett predikterat pris. Du kan se följande video för att lära dig mer om hur *Streamlit* fungerar: [https://www.youtube.com/watch?v=ggDa-RzPP7A&list=PLgzaMbMPEHEx9Als3F3sKKXexWnyEKH45&index=11&t=158s](https://www.youtube.com/watch?v=ggDa-RzPP7A&list=PLgzaMbMPEHEx9Als3F3sKKXexWnyEKH45&index=11&t=158s)

c) Det du gjort i denna uppgift, hade det kunnat användas i verkligheten? Till vad i sådana fall? </font>

Svar a): Nedan genomförs ett komplett ML-flöde för att prediktera `Price` utifrån bilens egenskaper. Vi börjar med en enkel baseline (linjär regression) och testar därefter om resultatet kan förbättras med en mer komplex modell (random forest).

In [10]:
import pandas as pd
import numpy as np

df = pd.read_csv('dataset/car_price_dataset.csv', sep=';')
df.head()

,Brand,Model,Year,Engine_Size,Fuel_Type,Transmission,Mileage,Doors,Owner_Count,Price
0,Kia,Rio,2020,4.2,Diesel,Manual,289944,3,5,8501
1,Chevrolet,Malibu,2012,2.0,Hybrid,Automatic,5356,2,3,12092
2,Mercedes,GLA,2020,4.2,Diesel,Automatic,231440,4,2,11171
3,Audi,Q5,2023,2.0,Electric,Manual,160971,2,1,11780
4,Volkswagen,Golf,2003,2.6,Hybrid,Semi-Automatic,286618,3,3,2867


In [11]:
# Ingen data saknas och samtliga kolumner har rimliga datatyper, så vi kan
# gå vidare direkt till modellering utan omfattande städning av datan.
print(df.isna().sum())
print(df.nunique())

Brand           0
Model           0
Year            0
Engine_Size     0
Fuel_Type       0
Transmission    0
Mileage         0
Doors           0
Owner_Count     0
Price           0
dtype: int64
Brand             10
Model             30
Year              24
Engine_Size       41
Fuel_Type          4
Transmission       3
Mileage         9854
Doors              4
Owner_Count        5
Price           6661
dtype: int64


In [12]:
from sklearn.model_selection import train_test_split

X = df.drop(columns='Price')
y = df['Price']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train.shape, X_test.shape

((8000, 9), (2000, 9))

In [13]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

categorical_features = ['Brand', 'Model', 'Fuel_Type', 'Transmission']
numerical_features = ['Year', 'Engine_Size', 'Mileage', 'Doors', 'Owner_Count']

preprocessor = ColumnTransformer([
    ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_features),
    ('num', StandardScaler(), numerical_features)
])

# Enkel baseline-modell
lin_pipeline = Pipeline([
    ('preprocessor', preprocessor),
    ('model', LinearRegression())
])

lin_pipeline.fit(X_train, y_train)
y_pred = lin_pipeline.predict(X_test)

rmse = mean_squared_error(y_test, y_pred) ** 0.5
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f'Linjär regression -> RMSE: {rmse:.2f}, MAE: {mae:.2f}, R2: {r2:.4f}')

Linjär regression -> RMSE: 64.91, MAE: 20.00, R2: 0.9995


In [14]:
from sklearn.ensemble import RandomForestRegressor

# Ett försök att förbättra resultatet med en mer komplex modell
rf_pipeline = Pipeline([
    ('preprocessor', preprocessor),
    ('model', RandomForestRegressor(n_estimators=300, random_state=42, n_jobs=-1))
])

rf_pipeline.fit(X_train, y_train)
y_pred_rf = rf_pipeline.predict(X_test)

rmse_rf = mean_squared_error(y_test, y_pred_rf) ** 0.5
mae_rf = mean_absolute_error(y_test, y_pred_rf)
r2_rf = r2_score(y_test, y_pred_rf)

print(f'Random forest -> RMSE: {rmse_rf:.2f}, MAE: {mae_rf:.2f}, R2: {r2_rf:.4f}')

Random forest -> RMSE: 327.93, MAE: 255.01, R2: 0.9883


In [15]:
import joblib

# Vi väljer den modell som presterade bäst på testdatan och sparar den så att
# Streamlit-appen (uppgift b) kan ladda in och använda den för prediktioner.
best_pipeline = lin_pipeline if r2 >= r2_rf else rf_pipeline
joblib.dump(best_pipeline, 'car_price_model.joblib')
print('Bästa modellen sparades till car_price_model.joblib')

Bästa modellen sparades till car_price_model.joblib


Svar b): En Streamlit-applikation har skapats i [car_price_app.py](car_price_app.py). Appen laddar den sparade modellen (`car_price_model.joblib`) och låter användaren mata in bilens uppgifter (märke, modell, årsmodell, motorstorlek, bränsletyp, växellåda, miltal, antal dörrar och antal tidigare ägare) via widgets i sidopanelen. När användaren klickar på knappen "Prediktera pris" visas modellens predikterade pris.

Appen startas via:
```bash
pip install streamlit
streamlit run car_price_app.py
```

Svar c): Ja, en modell av den här typen hade kunnat användas i verkligheten, exempelvis av:
- Bilhandlare och privatpersoner som snabbt vill få en indikation på ett rimligt pris vid köp eller försäljning av en bil.
- Marknadsplatser för begagnade bilar (t.ex. Blocket, Bytbil) som kan visa ett prisintervall till den som lägger upp en annons.
- Försäkringsbolag som behöver uppskatta ett fordons värde vid en skadereglering.

Man ska dock ha modellens begränsningar i åtanke: den är bara så bra som datan den tränats på. Om datasetet inte täcker alla marknader, tidsperioder eller specialfall (t.ex. skadade bilar eller ovanliga utrustningsnivåer) blir prediktionerna mindre tillförlitliga i verkligheten. Bilpriser påverkas dessutom av faktorer som inflation, ränteläge och utbud/efterfrågan över tid, vilket gör att modellen skulle behöva tränas om kontinuerligt för att hålla sig relevant.

<font color='orange'>16. I denna uppgift ska du göra ett komplett ML-flöde där du modellerar diamantpriser som finns tillgängliga i datasetet "diamonds.csv". Du kan läsa mer om datasetet här: [https://www.kaggle.com/datasets/shivam2503/diamonds](https://www.kaggle.com/datasets/shivam2503/diamonds)</font>

Svar: Nedan genomförs ett komplett ML-flöde för att prediktera `price` för diamanter. Datasetet innehåller 3 kategoriska variabler (`cut`, `color`, `clarity`) som alla har en naturlig rangordning samt 6 numeriska variabler som beskriver diamantens fysiska egenskaper (`carat`, `depth`, `table`, `x`, `y`, `z`).

In [16]:
import pandas as pd
import numpy as np

df = pd.read_csv('dataset/diamonds.csv', index_col=0)
df.head()

,carat,cut,color,clarity,depth,table,price,x,y,z
1,0.23,Ideal,E,SI2,61.5,55.0,326,3.95,3.98,2.43
2,0.21,Premium,E,SI1,59.8,61.0,326,3.89,3.84,2.31
3,0.23,Good,E,VS1,56.9,65.0,327,4.05,4.07,2.31
4,0.29,Premium,I,VS2,62.4,58.0,334,4.20,4.23,2.63
5,0.31,Good,J,SI2,63.3,58.0,335,4.34,4.35,2.75


In [17]:
print(df.isna().sum())
df.describe()

carat      0
cut        0
color      0
clarity    0
depth      0
table      0
price      0
x          0
y          0
z          0
dtype: int64


,carat,depth,table,price,x,y,z
count,53940.000000,53940.000000,53940.000000,53940.000000,53940.000000,53940.000000,53940.000000
mean,0.797940,61.749405,57.457184,3932.799722,5.731157,5.734526,3.538734
std,0.474011,1.432621,2.234491,3989.439738,1.121761,1.142135,0.705699
min,0.200000,43.000000,43.000000,326.000000,0.000000,0.000000,0.000000
25%,0.400000,61.000000,56.000000,950.000000,4.710000,4.720000,2.910000
50%,0.700000,61.800000,57.000000,2401.000000,5.700000,5.710000,3.530000
75%,1.040000,62.500000,59.000000,5324.250000,6.540000,6.540000,4.040000
max,5.010000,79.000000,95.000000,18823.000000,10.740000,58.900000,31.800000


Ingen data saknas, men `describe()` avslöjar att minimivärdet för `x`, `y` och `z` (diamantens dimensioner i mm) är 0, vilket är fysiskt omöjligt för en diamant med ett pris. En närmare koll visar även några rader där `y` eller `z` är orimligt stora (t.ex. 58.9 mm eller 31.8 mm), vilket ser ut att vara inmatningsfel (troligen felplacerat decimaltecken). Dessa rader städas bort innan modelleringen.

In [18]:
# Ta bort rader med omöjliga (0) eller orimliga (>15 mm) dimensioner
mask = (df.x == 0) | (df.y == 0) | (df.z == 0) | (df.x > 15) | (df.y > 15) | (df.z > 15)
print(f'Antal rader som städas bort: {mask.sum()}')

df = df[~mask].reset_index(drop=True)
df.shape

Antal rader som städas bort: 23


(53917, 10)

In [19]:
from sklearn.model_selection import train_test_split

X = df.drop(columns='price')
y = df['price']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train.shape, X_test.shape

((43133, 9), (10784, 9))

In [20]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OrdinalEncoder, StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

# cut, color och clarity är ordinala (har en inbördes rangordning), så vi
# använder OrdinalEncoder med rätt ordning istället för OneHotEncoder
cut_order = ['Fair', 'Good', 'Very Good', 'Premium', 'Ideal']
color_order = ['J', 'I', 'H', 'G', 'F', 'E', 'D']
clarity_order = ['I1', 'SI2', 'SI1', 'VS2', 'VS1', 'VVS2', 'VVS1', 'IF']

numerical_features = ['carat', 'depth', 'table', 'x', 'y', 'z']

preprocessor = ColumnTransformer([
    ('ordinal', OrdinalEncoder(categories=[cut_order, color_order, clarity_order]), ['cut', 'color', 'clarity']),
    ('num', StandardScaler(), numerical_features),
])

# Enkel baseline-modell
lin_pipeline = Pipeline([
    ('preprocessor', preprocessor),
    ('model', LinearRegression()),
])

lin_pipeline.fit(X_train, y_train)
y_pred = lin_pipeline.predict(X_test)

rmse = mean_squared_error(y_test, y_pred) ** 0.5
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f'Linjär regression -> RMSE: {rmse:.2f}, MAE: {mae:.2f}, R2: {r2:.4f}')

Linjär regression -> RMSE: 1174.45, MAE: 786.93, R2: 0.9125


In [21]:
from sklearn.ensemble import RandomForestRegressor

# Ett försök att förbättra resultatet med en mer komplex modell
rf_pipeline = Pipeline([
    ('preprocessor', preprocessor),
    ('model', RandomForestRegressor(n_estimators=300, random_state=42, n_jobs=-1)),
])

rf_pipeline.fit(X_train, y_train)
y_pred_rf = rf_pipeline.predict(X_test)

rmse_rf = mean_squared_error(y_test, y_pred_rf) ** 0.5
mae_rf = mean_absolute_error(y_test, y_pred_rf)
r2_rf = r2_score(y_test, y_pred_rf)

print(f'Random forest -> RMSE: {rmse_rf:.2f}, MAE: {mae_rf:.2f}, R2: {r2_rf:.4f}')

Random forest -> RMSE: 524.64, MAE: 260.32, R2: 0.9825


Random forest ger ett betydligt bättre resultat än den linjära regressionen (R² ≈ 0.98 jämfört med R² ≈ 0.91). Det är rimligt eftersom sambandet mellan diamanters egenskaper och pris inte är linjärt (t.ex. gör hopp i `carat`-vikt ofta ett oproportionerligt stort avtryck på priset), vilket en trädbaserad modell fångar bättre än en linjär modell.